opencog/reasoning/RuleEngine/rule-engine-src/pln/DefaultForwardChainerCB.cc

```diff
@@ -25,111 +25,175 @@
+//it was not working.Handles were not being passed.So I ended up using SchemeEval
+//with a hack (delete previously created SchemEval instance in order to use SchemeEval
+//with another atomspace)
 vector<Rule*> DefaultForwardChainerCB::choose_rule(FCMemory& fcmem)
 {
-	//create temporary atomspace and copy target
-	AtomSpace rule_atomspace;
-	Handle target = fcmem.get_cur_target();
-	if (target == Handle::UNDEFINED or NodeCast(target))
-		throw InvalidParamException(TRACE_INFO,
-				"Needs a target atom of type LINK");
-	as_->addAtom(target);
-	//create bindlink with target as an implicant
-	PLNCommons pc(&rule_atomspace);
-	Handle target_cpy = pc.replace_nodes_with_varnode(target, NODE);
-	Handle bind_link = pc.create_bindLink(target_cpy, false);
-	//copy rules to the temporary atomspace
-	vector<Rule*> rules = fcmem.get_rules();
-	for (Rule* r : rules)
-		as_->addAtom(r->get_handle());
-	//pattern match
-	DefaultImplicator imp(&rule_atomspace);
-	PatternMatch pm;
-	try {
-		pm.do_bindlink(bind_link, imp);
-	} catch (InvalidParamException& e) {
-		cout << "VALIDATION FAILED:" << endl << e.what() << endl;
-	}
-	//get matched bindLinks
-	auto matchings = imp.result_list;
-	if (matchings.empty())
-		return vector<Rule*> { };
-	HandleSeq bindlinks;
-	for (Handle hm : matchings) {
-		HandleSeq check_bindlink;
-		HandleSeq parents;
-		pc.get_top_level_parent(hm, parents);
-		for (Handle hp : parents) {
-			if (as_->getType(hp) == BIND_LINK
-					and find(bindlinks.begin(), bindlinks.end(), hp)
-							== bindlinks.end())
-				check_bindlink.push_back(hp);
-		}
-		//make sure matches are actually part of the premise list rather than the output of the bindLink
-		for (Handle hb : check_bindlink) {
-			auto outgoing = [this](Handle h) {return as_->getOutgoing(h);};
-			Handle hpremise = outgoing(outgoing(hb)[1])[0]; //extracting premise from (BindLink((ListLinK..)(ImpLink (premise) (..))))
-			if (pc.exists_in(hpremise, hm))
-				bindlinks.push_back(hb);
-		}
-	}
-	//transfer back bindlinks to main atomspace and  get their handle
-	HandleSeq copied_back;
-	for (Handle h : bindlinks)
-		copied_back.push_back(as_->addAtom(h));
-	//find the rules containing the bindLink in copied_back
-	vector<Rule*> matched_rules;
-	for (Rule* r : rules)
-		if (find(copied_back.begin(), copied_back.end(), r->get_handle())
-				!= copied_back.end())
-			matched_rules.push_back(r);
-	return matched_rules;
+    Handle target = fcmem.get_cur_target();
+    if (target == Handle::UNDEFINED or NodeCast(target))
+        throw InvalidParamException(TRACE_INFO,
+                                    "Needs a target atom of type LINK");
+    HandleSeq chosen_bindlinks;
+
+    if (LinkCast(target)) {
+        AtomSpace rule_atomspace;
+        SchemeEval* sceval = new SchemeEval(&rule_atomspace);
+        //Handle target_cpy=rule_atomspace.addAtom(target); xxx this doesn't work
+        Handle target_cpy = sceval->eval_h(SchemeSmob::to_string(target)); //xxx this works
```

Comment:
I just experimented: if I use `Handle target_cpy=rule_atomspace.addAtom(target);` instead of the string version, the test still passes, for me.

If I print out, using `target->toString()` and `target_cpy->toString()`, both variants print the same thing.

According to the atomspace design, both `Handle target_cpy = sceval->eval_h(SchemeSmob::to_string(target));` and `target_cpy=rule_atomspace.addAtom(target);` should result in exactly the same atomspace contents. If they are not the same, then there is a bug somewhere .. and thaat bug would need to be fixed.  The second form will, of course be much much faster than the first.